shared/fetcher.ts

```diff
@@ -0,0 +1,65 @@
+import axios from "axios";
+import Cookies from "js-cookie";
+
+import getConfig from "next/config";
+import { ITokenEnum } from "@shared/enum";
+
+const { publicRuntimeConfig } = getConfig();
+
+const baseURL = publicRuntimeConfig.BASE_URL;
+
+const api = axios.create({
+  baseURL,
+  timeout: 10000,
+  headers: { "content-type": "application/json"},
+});
+api.interceptors.request.use(
+  async (config) => {
+    const token = Cookies.get(ITokenEnum.token);
+    if (token) {
+      config.headers.Authorization = `Bearer ${token}`;
+    }
+    return config;
+  },
+  (error) => Promise.reject(error)
+);
+
+api.interceptors.response.use(
+  (response) => response,
+  async (error) => {
+    if (
+      error?.response?.status === 401 &&
+      error?.response?.config?.url !== "/users/refresh-token"
+    ) {
+      const token = Cookies.get(ITokenEnum.refreshToken);
+      const refreshApi = axios.create({
+        baseURL,
+        timeout: 3000,
+        headers: {
+          "content-type": "application/json",
+          authorization: `Bearer ${token}`,
+        },
+      });
+      return refreshApi.get("/users/refresh-token").then(
+        async (refreshResponse) => {
+          const data = refreshResponse?.data?.data;
+          if (data?.token) {
+            Cookies.set(ITokenEnum.token, data?.token);
+            Cookies.set(ITokenEnum.refreshToken, data?.refresh_token);
+            Cookies.set(ITokenEnum.anonymousToken, data?.anonymousToken);
+            error.config.headers.Authorization = `Bearer ${data?.token}`;
+            return axios.request(error?.config);
+          }
+          return Promise.reject(error);
+        },
+        (refreshError) => {
+          console.log("[refreshError]: ", refreshError);
+          return Promise.reject(error);
+        }
+      );
+    }
+    return Promise.reject(error);
+  }
+);
+
+export default api;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The code looks good overall, but there are a few areas that could be improved for better error handling and maintainability.

Firstly, in the request interceptor (lines 16-25), you're assuming that the token will always be available if it's not null. However, it would be safer to check if the token is not empty as well. This can help prevent sending requests with an empty Authorization header.

Secondly, in the response interceptor (lines 27-63), when the refresh token request fails (line 55), you're logging the error but not doing anything else. It would be beneficial to handle this error more gracefully, perhaps by redirecting the user to the login page or showing an error message.

Lastly, the axios instance `refreshApi` (lines 35-42) is created every time a request fails with a 401 status. This could lead to unnecessary overhead. Consider creating this instance outside of the interceptor and reusing it.

Here are the suggested changes:

```diff
api.interceptors.request.use(
  async (config) => {
    const token = Cookies.get(ITokenEnum.token);
-   if (token) {
+   if (token && token.trim() !== '') {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

...

-     const token = Cookies.get(ITokenEnum.refreshToken);
-     const refreshApi = axios.create({
-       baseURL,
-       timeout: 3000,
-       headers: {
-         "content-type": "application/json",
-         authorization: `Bearer ${token}`,
-       },
-     });
+     refreshApi.defaults.headers.authorization = `Bearer ${Cookies.get(ITokenEnum.refreshToken)}`;
      return refreshApi.get("/users/refresh-token").then(
        async (refreshResponse) => {
          const data = refreshResponse?.data?.data;
          if (data?.token) {
            Cookies.set(ITokenEnum.token, data?.token);
            Cookies.set(ITokenEnum.refreshToken, data?.refresh_token);
            Cookies.set(ITokenEnum.anonymousToken, data?.anonymousToken);
            error.config.headers.Authorization = `Bearer ${data?.token}`;
            return axios.request(error?.config);
          }
          return Promise.reject(error);
        },
-       (refreshError) => {
-         console.log("[refreshError]: ", refreshError);
+       (refreshError) => { 
+         // handle the error more gracefully here
          return Promise.reject(error);
        }
      );
    }
    return Promise.reject(error);
  }
);

...

// outside of the interceptor
const refreshApi = axios.create({
  baseURL,
  timeout: 3000,
  headers: {
    "content-type": "application/json",
  },
});
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->